src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,311 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
 
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
-    if (ucs_queue_is_empty(&ep->ext.stream->data)) {
+    rdesc = ucp_stream_recv_data_nb_internal(ep->ext.stream);
+    if (ucs_likely(rdesc != NULL)) {
+        *length = rdesc->length;
+        ret = ucp_stream_rdesc_data(rdesc);
+    } else {
         ret = UCS_STATUS_PTR(UCS_OK);
-        goto out;
     }
 
-    rdesc = ucs_queue_pull_elem_non_empty(&ep->ext.stream->data, ucp_recv_desc_t,
-                                          stream_queue);
-    *length = rdesc->length;
-    ret = ucp_stream_rdesc_data(rdesc);
-
-out:
     UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
 
     return ret;
 }
 
+static UCS_F_ALWAYS_INLINE void
+ucp_stream_rdesc_release(ucp_recv_desc_t *rdesc)
+{
+    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
```

Comment:
is it possible to unite some common code with tag-match ?

Reply:
No, until the rdesc_release logic is over complicated in tag-match and uses specific header sizes like  `ucp_stream_am_hdr_t` here.